{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import argparse\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parser = argparse.ArgumentParser(description='choose_genes_cv')\n",
    "# parser.add_argument('file_names', type=str, help='Name of folder and filenames for the promoters extracted')\n",
    "# parser.add_argument('promoter_bedfile', type=str, help='Input location of promoter bedfile')\n",
    "# parser.add_argument('Czechowski_rankedcv', type=str, help='Input location of Czechowski et al 2005 ranked cv dataset reanalysed by Will Nash')\n",
    "# parser.add_argument('Mergner_rankedcv', type=str, help='Input location of Mergner et al 2020 ranked cv dataset')\n",
    "# parser.add_argument('no_of_genes', type=int, help='Number of genes in each category to subset')\n",
    "# parser.add_argument('Czechowski_gene_categories', type=str, help='Output location of microarray gene category subsets')\n",
    "# parser.add_argument('Mergner_gene_categories', type=str, help='Output location of RNAseq gene category subsets')\n",
    "# parser.add_argument('promoter_mapped_motifs', type=str, help='Input location of promoter mapped motifs bed file')\n",
    "# parser.add_argument('promoters_filtered_contain_motifs', type=str, help='output location of the promoter bed file filtered so that each promoter contains at least one TFBS')\n",
    "# parser.add_argument('Czechowski_allgenes', type=str, help='Output location of all filtered microarray genes')\n",
    "# parser.add_argument('Mergner_allgenes', type=str, help='Output location of all filtered RNAseq genes')\n",
    "# parser.add_argument('promoters_gff3', type=str, help='Input location of promoters gff3 file')\n",
    "# args = parser.parse_args()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_proms_no_TFBS(promoter_bedfile, promoter_mapped_motifs,promoters_filtered_contain_motifs):\n",
    "    \"\"\"remove promoters which had no TFBSs found within them after filtering the FIMO output. Create output file of these\"\"\"\n",
    "    promoters = pd.read_table(promoter_bedfile, sep='\\t', header=None)\n",
    "    col = ['chr','start','stop','promoter_AGI','dot1', 'strand','source','type','dot2','attributes']\n",
    "    promoters.columns = col\n",
    "    mapped_motifs = pd.read_table(promoter_mapped_motifs, sep='\\t', header=None)\n",
    "    col2 = ['chr', 'start', 'stop', 'name_rep', 'score', 'strand', 'promoter_AGI', 'p-value', 'q-value', 'matched_sequence', 'TF_name', 'TF_family', 'TF_AGI']\n",
    "    mapped_motifs.columns = col2\n",
    "\n",
    "    merged = pd.merge(promoters,mapped_motifs, on='promoter_AGI', how='left',  suffixes=['', '_y'])\n",
    "    #remove NaNs in TF_AGI column\n",
    "    filtered_df = merged[merged.TF_AGI.notnull()]\n",
    "    #reduce columns\n",
    "    filtered_df = filtered_df[col]\n",
    "    #filter duplicates\n",
    "    idx = filtered_df.promoter_AGI.drop_duplicates().index\n",
    "    #this will return filtered df\n",
    "    no_dups = filtered_df.loc[idx,:]\n",
    "    no_dups.to_csv(promoters_filtered_contain_motifs, sep='\\t', header=None, index=False)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_only5UTR(promoter5UTR_bedfile, promoters_gff3):\n",
    "    \"\"\"remove genes where only the Araport11 5'UTR is present due to the promoter overlapping other genes\"\"\"\n",
    "    #read in df\n",
    "    promoter5UTR = pd.read_table(promoter5UTR_bedfile,sep='\\t', header=None)\n",
    "    col = ['chr','start','stop','AGI','dot1', 'strand','source','type','dot2','attributes']\n",
    "    promoter5UTR.columns = col\n",
    "    \n",
    "    # read in promoter only gff3\n",
    "    promoter_no_5UTR_df = pd.read_table(promoters_gff3, sep='\\t', header=None)\n",
    "    col = ['chr', 'source', 'type', 'start','stop', 'dot1','strand','dot2','attributes']\n",
    "    promoter_no_5UTR_df.columns = col\n",
    "    #add AGI column\n",
    "    promoter_no_5UTR_df_agi = promoter_no_5UTR_df.assign(AGI=promoter_no_5UTR_df.attributes.str.extract(r'ID=gene:(.*?)\\;'))\n",
    "    \n",
    "    #filter promoters in promoter5UTR but not in promoter_no_5UTR_df_agi\n",
    "    filtered = promoter5UTR[promoter5UTR.AGI.isin(promoter_no_5UTR_df_agi.AGI)]    \n",
    "    \n",
    "    #rename promoter5UTR_bedfile as including genes with only non-overlapping 5'UTRs\n",
    "    \n",
    "    oldextension = os.path.splitext(promoter5UTR_bedfile)[1]\n",
    "    oldname = os.path.splitext(promoter5UTR_bedfile)[0]\n",
    "    os.rename(promoter5UTR_bedfile, oldname + '_incl_only5UTR' + oldextension)\n",
    "    \n",
    "    #make a new file called the same name as promoter5UTR_bedfile\n",
    "    filtered.to_csv(promoter5UTR_bedfile,sep='\\t', header=None, index=False)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_genes_czechowski(promoter_bed, select_genes_file):\n",
    "    \"\"\"filter out genes from the microarray data which aren't in the promoter_bed\"\"\"\n",
    "    select_genes = pd.read_table(select_genes_file, sep='\\t', header=None)\n",
    "    cols = ['rank','probe_id','AGI','expression_mean','expression_SD','expression_CV','proportion_of_values_present_in_mas5','presence_in_araport11','constitutive_in_araport11']\n",
    "    select_genes.columns = cols\n",
    "    \n",
    "    promoters = pd.read_table(promoter_bed, sep='\\t', header=None)\n",
    "    col = ['chr','start','stop','AGI','dot1', 'strand','source','type','dot2','attributes']\n",
    "    promoters.columns = col\n",
    "\n",
    "    merged = pd.merge(promoters, select_genes, on='AGI', how='left')\n",
    "    #remove NaNs in expression_CV column\n",
    "    filtered_df = merged[merged.expression_CV.notnull()].copy()\n",
    "    #sort by CV value\n",
    "    filtered_df.sort_values('expression_CV', inplace=True, ignore_index=True)  \n",
    "    \n",
    "    #save df\n",
    "   # filtered_df.to_csv(Czechowski_allgenes,sep='\\t',header=None)\n",
    "    \n",
    "    return filtered_df   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_genes_mergner(promoter_bed, select_genes_file):\n",
    "    \"\"\"filter out genes from the RNA-seq data which aren't in the promoter_bed\"\"\"\n",
    "    select_genes = pd.read_csv(select_genes_file, header=0)\n",
    "    cols = ['AGI','transcription_class','transcription_family','expression_CV']\n",
    "    select_genes.columns = cols\n",
    "    #all present in Araport 11 column\n",
    "    select_genes['presence_in_araport11'] = 1\n",
    "    \n",
    "    return select_genes\n",
    "    promoters = pd.read_table(promoter_bed, sep='\\t', header=None)\n",
    "    col = ['chr','start','stop','AGI','dot1', 'strand','source','type','dot2','attributes']\n",
    "    promoters.columns = col\n",
    "\n",
    "    merged = pd.merge(promoters, select_genes, on='AGI', how='left')\n",
    "    #remove NaNs in expression_CV column\n",
    "    filtered_df = merged[merged.expression_CV.notnull()].copy()\n",
    "    #sort by CV value\n",
    "    filtered_df.sort_values('expression_CV', inplace=True, ignore_index=True)    \n",
    "    #save df\n",
    "  #  filtered_df.to_csv(Mergner_allgenes,sep='\\t',header=None)\n",
    "    \n",
    "    return filtered_df   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subSet_onCV(in_df, out_dir, no_of_genes):\n",
    "    '''\n",
    "    Extract the constitutive, variable, and control subsets based on CV values\n",
    "    '''\n",
    "    #filtering based on presence in the Araport11 annotation, define the first\n",
    "    #n rows as the constitutive set and add label\n",
    "    constitutive          = in_df[in_df.presence_in_araport11 == 1][0:no_of_genes]\n",
    "    constitutive['state'] = 'constitutive'\n",
    "\n",
    "    #define the last n rows as the variable set and add label\n",
    "    variable          = in_df[in_df.presence_in_araport11 == 1][-no_of_genes:]\n",
    "    variable['state'] = 'variable'\n",
    "    \n",
    "\n",
    "    #extract the rest of the rows as the control search space\n",
    "    mid_range    = in_df[in_df.presence_in_araport11 == 1][(no_of_genes+1):-(no_of_genes+1)]\n",
    "\n",
    "    #create 10 labelled bins\n",
    "    mid_range['bins'] = pd.Series(pd.qcut(mid_range['expression_CV'], q = 10, precision = 2))\n",
    "\n",
    "    #extract 10 random rows from these bins and label as the control set\n",
    "    samples_from_bins = mid_range.groupby('bins').apply(pd.DataFrame.sample, 10, random_state = 2)\n",
    "    samples_from_bins['state'] = 'control'\n",
    "    \n",
    "\n",
    "    #concatenate and write as output\n",
    "    output_set = pd.concat([constitutive[['AGI', 'state']], variable[['AGI', 'state']], samples_from_bins[['AGI', 'state']]], ignore_index = True)\n",
    "    return output_set\n",
    "   # output_set.to_csv(out_dir, sep = '\\t', index = False, header=False)\n",
    "    \n",
    "#     function from expressionVar_subsets_plot.py\n",
    "#     __author__ = \"Will Nash\"\n",
    "#     __copyright__ = \"Copyright 2020, The Earlham Institute\"\n",
    "#     __credits__ = [\"Will Nash\", \"Wilfried Haerty\"]\n",
    "#     __license__ = \"GPL\"\n",
    "#     __version__ = \"1.0\"\n",
    "#     __maintainer__ = \"Will Nash\"\n",
    "#     __email__ = \"will.nash@earlham.ac.uk\"\n",
    "#     __status__ = \"Testing\"\n",
    "#     __modified_by__ \"Sam Witham\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "file_names = 'non-overlapping_includingbidirectional_all_genes_newannotation'\n",
    "Czechowski_rankedcv = '../../data/genes/AtGE_dev_gcRMA__all_probes__CV.tsv'\n",
    "Mergner_rankedcv = '../../data/genes/RNA_CVs.csv'\n",
    "promoter_bedfile = f'../../data/output/{file_names}/FIMO/promoters_5UTR.bed'\n",
    "promoterno5UTR_bedfile = f'../../data/output/{file_names}/promoters.gff3'\n",
    "Czechowski_gene_categories = f'../../data/output/{file_names}/genes/promoters_5UTR_czechowski_constitutive_variable_random.txt'\n",
    "Mergner_gene_categories = f'../../data/output/{file_names}/genes/promoters_5UTR_mergner_constitutive_variable_random.txt'\n",
    "no_of_genes = 100\n",
    "promoter_mapped_motifs = f'../../data/output/{file_names}/FIMO/promoters_5UTR_motifs_mapped.bed'\n",
    "promoters_filtered_contain_motifs = f'../../data/output/{file_names}/FIMO/promoters_5UTR_filtered_contain_motifs.bed'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory  ../../data/output/non-overlapping_includingbidirectional_all_genes_newannotation/genes  already exists\n"
     ]
    }
   ],
   "source": [
    "#make directory for the output files to be exported to\n",
    "#dirName = f'{args.directory_path}/data/output/{args.file_names}'\n",
    "dirName = f'../../data/output/{file_names}/genes'\n",
    "try:\n",
    "    # Create target Directory\n",
    "    os.mkdir(dirName)\n",
    "    print(\"Directory \" , dirName ,  \" created\") \n",
    "except FileExistsError:\n",
    "    print(\"Directory \" , dirName ,  \" already exists\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "remove_proms_no_TFBS(promoter_bedfile, promoter_mapped_motifs,promoters_filtered_contain_motifs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "remove_only5UTR(promoter_bedfile, promoterno5UTR_bedfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_czechowski = filter_genes_czechowski(promoters_filtered_contain_motifs,Czechowski_rankedcv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_mergner = filter_genes_mergner(promoters_filtered_contain_motifs,Mergner_rankedcv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AGI</th>\n",
       "      <th>transcription_class</th>\n",
       "      <th>transcription_family</th>\n",
       "      <th>expression_CV</th>\n",
       "      <th>presence_in_araport11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G79940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.181119</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT2G32170</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.192912</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT3G50860</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.209712</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT3G49850</td>\n",
       "      <td>TF</td>\n",
       "      <td>MYB-related</td>\n",
       "      <td>0.211639</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT5G25760</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.213157</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25152</th>\n",
       "      <td>AT2G25685</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.385165</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25153</th>\n",
       "      <td>AT4G29090</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.385165</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25154</th>\n",
       "      <td>AT3G59170</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.385165</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25155</th>\n",
       "      <td>AT1G33730</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.385165</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25156</th>\n",
       "      <td>AT3G27999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.385165</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25157 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             AGI transcription_class transcription_family  expression_CV  \\\n",
       "0      AT1G79940                 NaN                  NaN       0.181119   \n",
       "1      AT2G32170                 NaN                  NaN       0.192912   \n",
       "2      AT3G50860                 NaN                  NaN       0.209712   \n",
       "3      AT3G49850                  TF          MYB-related       0.211639   \n",
       "4      AT5G25760                 NaN                  NaN       0.213157   \n",
       "...          ...                 ...                  ...            ...   \n",
       "25152  AT2G25685                 NaN                  NaN       5.385165   \n",
       "25153  AT4G29090                 NaN                  NaN       5.385165   \n",
       "25154  AT3G59170                 NaN                  NaN       5.385165   \n",
       "25155  AT1G33730                 NaN                  NaN       5.385165   \n",
       "25156  AT3G27999                 NaN                  NaN       5.385165   \n",
       "\n",
       "       presence_in_araport11  \n",
       "0                          1  \n",
       "1                          1  \n",
       "2                          1  \n",
       "3                          1  \n",
       "4                          1  \n",
       "...                      ...  \n",
       "25152                      1  \n",
       "25153                      1  \n",
       "25154                      1  \n",
       "25155                      1  \n",
       "25156                      1  \n",
       "\n",
       "[25157 rows x 5 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_mergner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AGI</th>\n",
       "      <th>state</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G34270</td>\n",
       "      <td>constitutive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT4G33380</td>\n",
       "      <td>constitutive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G13320</td>\n",
       "      <td>constitutive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT2G28390</td>\n",
       "      <td>constitutive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G71820</td>\n",
       "      <td>constitutive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>AT5G02100</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>AT2G27830</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>AT4G27652</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>AT5G65710</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>AT4G39950</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           AGI         state\n",
       "0    AT4G34270  constitutive\n",
       "1    AT4G33380  constitutive\n",
       "2    AT1G13320  constitutive\n",
       "3    AT2G28390  constitutive\n",
       "4    AT1G71820  constitutive\n",
       "..         ...           ...\n",
       "295  AT5G02100       control\n",
       "296  AT2G27830       control\n",
       "297  AT4G27652       control\n",
       "298  AT5G65710       control\n",
       "299  AT4G39950       control\n",
       "\n",
       "[300 rows x 2 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#czechowksi subset\n",
    "subSet_onCV(filtered_czechowski,Czechowski_gene_categories,no_of_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AGI</th>\n",
       "      <th>state</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G79940</td>\n",
       "      <td>constitutive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT2G32170</td>\n",
       "      <td>constitutive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT3G50860</td>\n",
       "      <td>constitutive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT3G49850</td>\n",
       "      <td>constitutive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT5G25760</td>\n",
       "      <td>constitutive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>AT2G33670</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>AT4G31230</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>AT3G26250</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>AT3G27785</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>AT1G19060</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           AGI         state\n",
       "0    AT1G79940  constitutive\n",
       "1    AT2G32170  constitutive\n",
       "2    AT3G50860  constitutive\n",
       "3    AT3G49850  constitutive\n",
       "4    AT5G25760  constitutive\n",
       "..         ...           ...\n",
       "295  AT2G33670       control\n",
       "296  AT4G31230       control\n",
       "297  AT3G26250       control\n",
       "298  AT3G27785       control\n",
       "299  AT1G19060       control\n",
       "\n",
       "[300 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#mergner subset\n",
    "subSet_onCV(filtered_mergner,Mergner_gene_categories,no_of_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:PromoterArchitecturePipeline] *",
   "language": "python",
   "name": "conda-env-PromoterArchitecturePipeline-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
