{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "#import argparse\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parser = argparse.ArgumentParser(description='choose_genes_cv')\n",
    "# parser.add_argument('file_names', type=str, help='Name of folder and filenames for the promoters extracted')\n",
    "# parser.add_argument('promoter_bedfile', type=str, help='Input location of promoter bedfile')\n",
    "# parser.add_argument('Czechowski_rankedcv', type=str, help='Input location of Czechowski et al 2005 ranked cv dataset reanalysed by Will Nash')\n",
    "# parser.add_argument('Mergner_rankedcv', type=str, help='Input location of Mergner et al 2020 ranked cv dataset')\n",
    "# parser.add_argument('no_of_genes', type=int, help='Number of genes in each category to subset')\n",
    "# parser.add_argument('Czechowski_gene_categories', type=str, help='Output location of microarray gene category subsets')\n",
    "# parser.add_argument('Mergner_gene_categories', type=str, help='Output location of RNAseq gene category subsets')\n",
    "# parser.add_argument('promoter_mapped_motifs', type=str, help='Input location of promoter mapped motifs bed file')\n",
    "# parser.add_argument('promoters_filtered_contain_motifs', type=str, help='output location of the promoter bed file filtered so that each promoter contains at least one TFBS')\n",
    "# parser.add_argument('Czechowski_allgenes', type=str, help='Output location of all filtered microarray genes')\n",
    "# parser.add_argument('Mergner_allgenes', type=str, help='Output location of all filtered RNAseq genes')\n",
    "# parser.add_argument('promoters_gff3', type=str, help='Input location of promoters gff3 file')\n",
    "# parser.add_argument('option_remove_only5UTR', type=str, help='Optionally remove genes where only the Araport11 5UTR is present due to the promoter overlapping other genes',default = True, nargs=\"?\")\n",
    "# args = parser.parse_args()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "file_names = 'non-overlapping_includingbidirectional_all_genes_newannotation'\n",
    "#file_names = 'arabidopsis_selected_tissues'\n",
    "#Czechowski_rankedcv = '../../data/genes/AtGE_dev_gcRMA__all_probes__CV.tsv'\n",
    "Czechowski_rankedcv = f'../../data/output/{file_names}/genes/filtered80/AtGE_dev_gcRMA__all_probes__CV.tsv'\n",
    "#Mergner_rankedcv = '../../data/genes/RNA_CVs.csv'\n",
    "#promoter_bedfile = f'../../data/output/{file_names}/FIMO/promoters_5UTR.bed'\n",
    "#promoterno5UTR_bedfile = f'../../data/output/{file_names}/promoters.gff3'\n",
    "#Czechowski_gene_categories = f'../../data/output/{file_names}/genes/promoters_5UTR_czechowski_constitutive_variable_random_300.txt'\n",
    "#Mergner_gene_categories = f'../../data/output/{file_names}/genes/promoters_5UTR_mergner_constitutive_variable_random_300.txt'\n",
    "no_of_genes = 100\n",
    "#promoter_mapped_motifs = f'../../data/output/{file_names}/FIMO/promoters_5UTR_motifs_mapped.bed'\n",
    "#promoters_filtered_contain_motifs = f'../../data/output/{file_names}/FIMO/promoters_5UTR_filtered_contain_motifs.bed'\n",
    "TF_list = '../../data/genes/Ath_TF_list.txt'\n",
    "#Czechowski_allgenes = f'../../data/output/{file_names}/genes/promoters_5UTR_czechowski_allfilteredgenes.txt'\n",
    "\n",
    "Czechowski_gene_categories_tfs = f'../../data/output/{file_names}/genes/tfs_czechowski_constitutive_variable_random_300'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_genes_czechowski(TF_list, select_genes_file):\n",
    "    \"\"\"filter out genes from the microarray data which aren't in the promoter_bed\"\"\"\n",
    "    #select_genes = pd.read_table(select_genes_file, sep='\\t', header=0)\n",
    "    select_genes = pd.read_table(select_genes_file, sep='\\t', header=None)\n",
    "    cols = ['rank','probe_id','AGI','expression_mean','expression_SD','expression_CV','proportion_of_values_present_in_mas5','presence_in_araport11','constitutive_in_araport11']\n",
    "    select_genes.columns = cols\n",
    "    #make AGI uppercase\n",
    "    select_genes.AGI = select_genes.AGI.str.upper()\n",
    "\n",
    "    #read in TF list\n",
    "    TF_list_df = pd.read_table(TF_list, sep='\\t',header=1)\n",
    "    \n",
    "    merged = pd.merge(TF_list_df, select_genes, left_on='Gene_ID',right_on='AGI', how='left')\n",
    "    #remove NaNs in expression_CV column\n",
    "    filtered_df = merged[merged.expression_CV.notnull()].copy()\n",
    "    \n",
    "    #sort by TF ID\n",
    "    filtered_df.sort_values(['TF_ID'], inplace=True, ignore_index=True)\n",
    "   \n",
    "    #remove duplicates based on Gene_ID column\n",
    "    filtered_df.drop_duplicates(subset=['Gene_ID'],inplace=True)\n",
    "    \n",
    "    #save df\n",
    "    #filtered_df.to_csv(args.Czechowski_allgenes,sep='\\t',columns=filtered_df.columns, index=False)\n",
    "    \n",
    "    #sort by CV value\n",
    "    filtered_df.sort_values('expression_CV', inplace=True, ignore_index=True) \n",
    "    \n",
    "    return filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def filter_genes_mergner(promoter_bed, select_genes_file):\n",
    "#     \"\"\"filter out genes from the RNA-seq data which aren't in the promoter_bed\"\"\"\n",
    "#     select_genes = pd.read_csv(select_genes_file, header=0)\n",
    "#     cols = ['AGI','transcription_class','transcription_family','expression_CV']\n",
    "#     select_genes.columns = cols\n",
    "#     #all present in Araport 11 column\n",
    "#     select_genes['presence_in_araport11'] = 1\n",
    "    \n",
    "#     promoters = pd.read_table(promoter_bed, sep='\\t', header=None)\n",
    "#     col = ['chr','start','stop','AGI','dot1', 'strand','source','type','dot2','attributes']\n",
    "#     promoters.columns = col\n",
    "\n",
    "#     merged = pd.merge(promoters, select_genes, on='AGI', how='left')\n",
    "#     #remove NaNs in expression_CV column\n",
    "#     filtered_df = merged[merged.expression_CV.notnull()].copy()\n",
    "    \n",
    "#     #sort by chr and start\n",
    "#     filtered_df.sort_values(['chr','start'], inplace=True, ignore_index=True)\n",
    "    \n",
    "#     #save df\n",
    "#     filtered_df.to_csv(args.Mergner_allgenes,sep='\\t',columns=filtered_df.columns,index=False)\n",
    "    \n",
    "#     #sort by CV value\n",
    "#     filtered_df.sort_values('expression_CV', inplace=True, ignore_index=True)\n",
    "#     return filtered_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subSet_onCV(in_df, out_dir, no_of_genes):\n",
    "    '''\n",
    "    Extract the constitutive, variable, and control subsets based on CV values\n",
    "    '''\n",
    "    #filtering based on presence in the Araport11 annotation, define the first\n",
    "    #n rows as the constitutive set and add label\n",
    "    constitutive          = in_df[in_df.presence_in_araport11 == 1][0:no_of_genes]\n",
    "    constitutive['state'] = 'constitutive'\n",
    "\n",
    "    #define the last n rows as the variable set and add label\n",
    "    variable          = in_df[in_df.presence_in_araport11 == 1][-no_of_genes:]\n",
    "    variable['state'] = 'variable'\n",
    "\n",
    "    #extract the rest of the rows as the control search space\n",
    "    mid_range    = in_df[in_df.presence_in_araport11 == 1][(no_of_genes+1):-(no_of_genes+1)]\n",
    "\n",
    "    #create 10 labelled bins\n",
    "    mid_range['bins'] = pd.Series(pd.qcut(mid_range['expression_CV'], q = 10, precision = 2))\n",
    "\n",
    "    #extract 10 random rows from these bins and label as the control set\n",
    "    sample = no_of_genes/10\n",
    "    sample_integar = int(str(sample).replace('.0', '')) #convert sample to an integar\n",
    "    samples_from_bins = mid_range.groupby('bins').apply(pd.DataFrame.sample, sample_integar, random_state = 2)\n",
    "    samples_from_bins['state'] = 'control'\n",
    "\n",
    "    #concatenate and write as output\n",
    "    output_set = pd.concat([constitutive[['AGI', 'state']], variable[['AGI', 'state']], samples_from_bins[['AGI', 'state']]], ignore_index = True)\n",
    "    output_set.to_csv(out_dir, sep = '\\t', index = False, header=False)\n",
    "    \n",
    "    #function from expressionVar_subsets_plot.py\n",
    "    #__author__ = \"Will Nash\"\n",
    "    # __copyright__ = \"Copyright 2020, The Earlham Institute\"\n",
    "    # __credits__ = [\"Will Nash\", \"Wilfried Haerty\"]\n",
    "    # __license__ = \"GPL\"\n",
    "    # __version__ = \"1.0\"\n",
    "    # __maintainer__ = \"Will Nash\"\n",
    "    # __email__ = \"will.nash@earlham.ac.uk\"\n",
    "    # __status__ = \"Testing\"\n",
    "    #__modified_by__ \"Sam Witham\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory  ../../data/output/non-overlapping_includingbidirectional_all_genes_newannotation/genes  already exists\n"
     ]
    }
   ],
   "source": [
    "#make directory for the output files to be exported to\n",
    "#dirName = f'{args.directory_path}/data/output/{args.file_names}'\n",
    "dirName = f'../../data/output/{file_names}/genes'\n",
    "try:\n",
    "    # Create target Directory\n",
    "    os.mkdir(dirName)\n",
    "    print(\"Directory \" , dirName ,  \" created\") \n",
    "except FileExistsError:\n",
    "    print(\"Directory \" , dirName ,  \" already exists\")\n",
    "\n",
    "\n",
    "filtered_czechowski = filter_genes_czechowski(TF_list,Czechowski_rankedcv)\n",
    "#filtered_mergner = filter_genes_mergner(args.promoters_filtered_contain_motifs,args.Mergner_rankedcv)\n",
    "#czechowksi subset\n",
    "subSet_onCV(filtered_czechowski,Czechowski_gene_categories_tfs,no_of_genes)\n",
    "#mergner subset\n",
    "#subSet_onCV(filtered_mergner,args.Mergner_gene_categories,args.no_of_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #make directory for the output files to be exported to\n",
    "# #dirName = f'{args.directory_path}/data/output/{args.file_names}'\n",
    "# dirName = f'../../data/output/{args.file_names}/genes'\n",
    "# try:\n",
    "#     # Create target Directory\n",
    "#     os.mkdir(dirName)\n",
    "#     print(\"Directory \" , dirName ,  \" created\") \n",
    "# except FileExistsError:\n",
    "#     print(\"Directory \" , dirName ,  \" already exists\")\n",
    "\n",
    "\n",
    "# filtered_czechowski = filter_genes_czechowski(args.promoters_filtered_contain_motifs,args.Czechowski_rankedcv)\n",
    "# #filtered_mergner = filter_genes_mergner(args.promoters_filtered_contain_motifs,args.Mergner_rankedcv)\n",
    "# #czechowksi subset\n",
    "# subSet_onCV(filtered_czechowski,args.Czechowski_gene_categories,args.no_of_genes)\n",
    "# #mergner subset\n",
    "# #subSet_onCV(filtered_mergner,args.Mergner_gene_categories,args.no_of_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF_ID</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Family</th>\n",
       "      <th>rank</th>\n",
       "      <th>probe_id</th>\n",
       "      <th>AGI</th>\n",
       "      <th>expression_mean</th>\n",
       "      <th>expression_SD</th>\n",
       "      <th>expression_CV</th>\n",
       "      <th>proportion_of_values_present_in_mas5</th>\n",
       "      <th>presence_in_araport11</th>\n",
       "      <th>constitutive_in_araport11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G26640.1</td>\n",
       "      <td>AT4G26640</td>\n",
       "      <td>WRKY</td>\n",
       "      <td>62.0</td>\n",
       "      <td>253983_at</td>\n",
       "      <td>AT4G26640</td>\n",
       "      <td>249.301390</td>\n",
       "      <td>47.634972</td>\n",
       "      <td>0.191074</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT5G45050.1</td>\n",
       "      <td>AT5G45050</td>\n",
       "      <td>WRKY</td>\n",
       "      <td>169.0</td>\n",
       "      <td>248973_at</td>\n",
       "      <td>AT5G45050</td>\n",
       "      <td>120.863167</td>\n",
       "      <td>25.667745</td>\n",
       "      <td>0.212370</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT5G09410.1</td>\n",
       "      <td>AT5G09410</td>\n",
       "      <td>CAMTA</td>\n",
       "      <td>176.0</td>\n",
       "      <td>245910_at</td>\n",
       "      <td>AT5G09410</td>\n",
       "      <td>389.393405</td>\n",
       "      <td>82.922480</td>\n",
       "      <td>0.212953</td>\n",
       "      <td>99.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT5G67580.1</td>\n",
       "      <td>AT5G67580</td>\n",
       "      <td>MYB_related</td>\n",
       "      <td>206.0</td>\n",
       "      <td>247012_at</td>\n",
       "      <td>AT5G67580</td>\n",
       "      <td>61.548600</td>\n",
       "      <td>13.354799</td>\n",
       "      <td>0.216980</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT4G38900.1</td>\n",
       "      <td>AT4G38900</td>\n",
       "      <td>bZIP</td>\n",
       "      <td>308.0</td>\n",
       "      <td>252969_at</td>\n",
       "      <td>AT4G38900</td>\n",
       "      <td>152.627983</td>\n",
       "      <td>34.909329</td>\n",
       "      <td>0.228722</td>\n",
       "      <td>99.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>AT3G16350.1</td>\n",
       "      <td>AT3G16350</td>\n",
       "      <td>MYB_related</td>\n",
       "      <td>11956.0</td>\n",
       "      <td>259379_at</td>\n",
       "      <td>AT3G16350</td>\n",
       "      <td>82.968393</td>\n",
       "      <td>242.133141</td>\n",
       "      <td>2.918378</td>\n",
       "      <td>81.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>AT2G40620.1</td>\n",
       "      <td>AT2G40620</td>\n",
       "      <td>bZIP</td>\n",
       "      <td>11963.0</td>\n",
       "      <td>255878_at</td>\n",
       "      <td>AT2G40620</td>\n",
       "      <td>123.417440</td>\n",
       "      <td>369.556407</td>\n",
       "      <td>2.994361</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>AT5G16540.1</td>\n",
       "      <td>AT5G16540</td>\n",
       "      <td>C3H</td>\n",
       "      <td>11972.0</td>\n",
       "      <td>250128_at</td>\n",
       "      <td>AT5G16540</td>\n",
       "      <td>357.916799</td>\n",
       "      <td>1102.500621</td>\n",
       "      <td>3.080327</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>AT1G26610.1</td>\n",
       "      <td>AT1G26610</td>\n",
       "      <td>C2H2</td>\n",
       "      <td>11989.0</td>\n",
       "      <td>260997_at</td>\n",
       "      <td>AT1G26610</td>\n",
       "      <td>30.842648</td>\n",
       "      <td>99.945493</td>\n",
       "      <td>3.240497</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569</th>\n",
       "      <td>AT3G57390.1</td>\n",
       "      <td>AT3G57390</td>\n",
       "      <td>MIKC_MADS</td>\n",
       "      <td>11993.0</td>\n",
       "      <td>251623_at</td>\n",
       "      <td>AT3G57390</td>\n",
       "      <td>182.605145</td>\n",
       "      <td>603.883365</td>\n",
       "      <td>3.307045</td>\n",
       "      <td>99.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>570 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           TF_ID    Gene_ID       Family     rank   probe_id        AGI  \\\n",
       "0    AT4G26640.1  AT4G26640         WRKY     62.0  253983_at  AT4G26640   \n",
       "1    AT5G45050.1  AT5G45050         WRKY    169.0  248973_at  AT5G45050   \n",
       "2    AT5G09410.1  AT5G09410        CAMTA    176.0  245910_at  AT5G09410   \n",
       "3    AT5G67580.1  AT5G67580  MYB_related    206.0  247012_at  AT5G67580   \n",
       "4    AT4G38900.1  AT4G38900         bZIP    308.0  252969_at  AT4G38900   \n",
       "..           ...        ...          ...      ...        ...        ...   \n",
       "565  AT3G16350.1  AT3G16350  MYB_related  11956.0  259379_at  AT3G16350   \n",
       "566  AT2G40620.1  AT2G40620         bZIP  11963.0  255878_at  AT2G40620   \n",
       "567  AT5G16540.1  AT5G16540          C3H  11972.0  250128_at  AT5G16540   \n",
       "568  AT1G26610.1  AT1G26610         C2H2  11989.0  260997_at  AT1G26610   \n",
       "569  AT3G57390.1  AT3G57390    MIKC_MADS  11993.0  251623_at  AT3G57390   \n",
       "\n",
       "     expression_mean  expression_SD  expression_CV  \\\n",
       "0         249.301390      47.634972       0.191074   \n",
       "1         120.863167      25.667745       0.212370   \n",
       "2         389.393405      82.922480       0.212953   \n",
       "3          61.548600      13.354799       0.216980   \n",
       "4         152.627983      34.909329       0.228722   \n",
       "..               ...            ...            ...   \n",
       "565        82.968393     242.133141       2.918378   \n",
       "566       123.417440     369.556407       2.994361   \n",
       "567       357.916799    1102.500621       3.080327   \n",
       "568        30.842648      99.945493       3.240497   \n",
       "569       182.605145     603.883365       3.307045   \n",
       "\n",
       "     proportion_of_values_present_in_mas5  presence_in_araport11  \\\n",
       "0                                    98.0                    1.0   \n",
       "1                                    98.0                    1.0   \n",
       "2                                    99.0                    1.0   \n",
       "3                                   100.0                    1.0   \n",
       "4                                    99.0                    1.0   \n",
       "..                                    ...                    ...   \n",
       "565                                  81.0                    1.0   \n",
       "566                                 100.0                    1.0   \n",
       "567                                  98.0                    1.0   \n",
       "568                                  98.0                    1.0   \n",
       "569                                  99.0                    1.0   \n",
       "\n",
       "     constitutive_in_araport11  \n",
       "0                          0.0  \n",
       "1                          0.0  \n",
       "2                          1.0  \n",
       "3                          0.0  \n",
       "4                          0.0  \n",
       "..                         ...  \n",
       "565                        0.0  \n",
       "566                        0.0  \n",
       "567                        0.0  \n",
       "568                        0.0  \n",
       "569                        0.0  \n",
       "\n",
       "[570 rows x 12 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_czechowski"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:PromoterArchitecturePipeline] *",
   "language": "python",
   "name": "conda-env-PromoterArchitecturePipeline-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
