{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import argparse\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = argparse.ArgumentParser(description='choose_genes_cv')\n",
    "parser.add_argument('file_names', type=str, help='Name of folder and filenames for the promoters extracted')\n",
    "parser.add_argument('promoter_bedfile', type=str, help='Input location of promoter bedfile')\n",
    "parser.add_argument('Czechowski_rankedcv', type=str, help='Input location of Czechowski et al 2005 ranked cv dataset reanalysed by Will Nash')\n",
    "parser.add_argument('Mergner_rankedcv', type=str, help='Input location of Mergner et al 2020 ranked cv dataset')\n",
    "parser.add_argument('no_of_genes', type=int, help='Number of genes in each category to subset')\n",
    "parser.add_argument('Czechowski_gene_categories', type=str, help='Output location of microarray gene category subsets')\n",
    "parser.add_argument('Mergner_gene_categories', type=str, help='Output location of RNAseq gene category subsets')\n",
    "parser.add_argument('promoter_mapped_motifs', type=str, help='Input location of promoter mapped motifs bed file')\n",
    "parser.add_argument('promoters_filtered_contain_motifs', type=str, help='output location of the promoter bed file filtered so that each promoter contains at least one TFBS')\n",
    "parser.add_argument('Czechowski_allgenes', type=str, help='Output location of all filtered microarray genes')\n",
    "parser.add_argument('Mergner_allgenes', type=str, help='Output location of all filtered RNAseq genes')\n",
    "\n",
    "args = parser.parse_args()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_proms_no_TFBS(promoter_bedfile, promoter_mapped_motifs,promoters_filtered_contain_motifs):\n",
    "    \"\"\"remove promoters which had no TFBSs found within them after filtering the FIMO output. Create output file of these\"\"\"\n",
    "    promoters = pd.read_table(promoter_bedfile, sep='\\t', header=None)\n",
    "    col = ['chr','start','stop','promoter_AGI','dot1', 'strand','source','type','dot2','attributes']\n",
    "    promoters.columns = col\n",
    "    mapped_motifs = pd.read_table(promoter_mapped_motifs, sep='\\t', header=None)\n",
    "    col2 = ['chr', 'start', 'stop', 'name_rep', 'score', 'strand', 'promoter_AGI', 'p-value', 'q-value', 'matched_sequence', 'TF_name', 'TF_family', 'TF_AGI']\n",
    "    mapped_motifs.columns = col2\n",
    "\n",
    "    merged = pd.merge(promoters,mapped_motifs, on='promoter_AGI', how='left',  suffixes=['', '_y'])\n",
    "    #remove NaNs in TF_AGI column\n",
    "    filtered_df = merged[merged.TF_AGI.notnull()]\n",
    "    #reduce columns\n",
    "    filtered_df = filtered_df[col]\n",
    "    #filter duplicates\n",
    "    idx = filtered_df.promoter_AGI.drop_duplicates().index\n",
    "    #this will return filtered df\n",
    "    no_dups = filtered_df.loc[idx,:]\n",
    "    no_dups.to_csv(promoters_filtered_contain_motifs, sep='\\t', header=None, index=False)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_genes_czechowski(promoter_bed, select_genes_file):\n",
    "    \"\"\"filter out genes from the microarray data which aren't in the promoter_bed\"\"\"\n",
    "    select_genes = pd.read_table(select_genes_file, sep='\\t', header=None)\n",
    "    cols = ['rank','probe_id','AGI','expression_mean','expression_SD','expression_CV','proportion_of_values_present_in_mas5','presence_in_araport11','constitutive_in_araport11']\n",
    "    select_genes.columns = cols\n",
    "    \n",
    "    promoters = pd.read_table(promoter_bed, sep='\\t', header=None)\n",
    "    col = ['chr','start','stop','AGI','dot1', 'strand','source','type','dot2','attributes']\n",
    "    promoters.columns = col\n",
    "\n",
    "    merged = pd.merge(promoters, select_genes, on='AGI', how='left')\n",
    "    #remove NaNs in expression_CV column\n",
    "    filtered_df = merged[merged.expression_CV.notnull()]\n",
    "    #save df\n",
    "    filtered_df.to_csv(Czechowski_allgenes,sep='\\t',header=None)\n",
    "    \n",
    "    return filtered_df   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_genes_mergner(promoter_bed, select_genes_file):\n",
    "    \"\"\"filter out genes from the RNA-seq data which aren't in the promoter_bed\"\"\"\n",
    "    select_genes = pd.read_csv(select_genes_file, header=0)\n",
    "    cols = ['AGI','transcription_class','transcription_family','expression_CV']\n",
    "    select_genes.columns = cols\n",
    "    #all present in Araport 11 column\n",
    "    select_genes['presence_in_araport11'] = 1\n",
    "    \n",
    "    promoters = pd.read_table(promoter_bed, sep='\\t', header=None)\n",
    "    col = ['chr','start','stop','AGI','dot1', 'strand','source','type','dot2','attributes']\n",
    "    promoters.columns = col\n",
    "\n",
    "    merged = pd.merge(promoters, select_genes, on='AGI', how='left')\n",
    "    #remove NaNs in expression_CV column\n",
    "    filtered_df = merged[merged.expression_CV.notnull()]\n",
    "    #save df\n",
    "    filtered_df.to_csv(Mergner_allgenes,sep='\\t',header=None)\n",
    "    \n",
    "    return filtered_df   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subSet_onCV(in_df, out_dir, no_of_genes):\n",
    "    '''\n",
    "    Extract the constitutive, variable, and control subsets based on CV values\n",
    "    '''\n",
    "    #fliltering based on presence in the Araport11 annotation, define the first\n",
    "    #n rows as the constitutive set and add label\n",
    "    constitutive          = in_df[in_df.presence_in_araport11 == 1][0:no_of_genes]\n",
    "    constitutive['state'] = 'constitutive'\n",
    "\n",
    "    #define the last n rows as the variable set and add label\n",
    "    variable          = in_df[in_df.presence_in_araport11 == 1][-no_of_genes:]\n",
    "    variable['state'] = 'variable'\n",
    "\n",
    "    #extract the rest of the rows as the control search space\n",
    "    mid_range    = in_df[in_df.presence_in_araport11 == 1][(no_of_genes+1):-(no_of_genes+1)]\n",
    "\n",
    "    #create 10 labelled bins\n",
    "    mid_range['bins'] = pd.Series(pd.qcut(mid_range['expression_CV'], q = 10, precision = 2))\n",
    "\n",
    "    #extract 10 random rows from these bins and label as the control set\n",
    "    samples_from_bins = mid_range.groupby('bins').apply(pd.DataFrame.sample, 10, random_state = 2)\n",
    "    samples_from_bins['state'] = 'control'\n",
    "\n",
    "    #concatenate and write as output\n",
    "    output_set = pd.concat([constitutive[['AGI', 'state']], variable[['AGI', 'state']], samples_from_bins[['AGI', 'state']]], ignore_index = True)\n",
    "    output_set.to_csv(out_dir, sep = '\\t', index = False, header=False)\n",
    "    \n",
    "    #function from expressionVar_subsets_plot.py\n",
    "    #__author__ = \"Will Nash\"\n",
    "    # __copyright__ = \"Copyright 2020, The Earlham Institute\"\n",
    "    # __credits__ = [\"Will Nash\", \"Wilfried Haerty\"]\n",
    "    # __license__ = \"GPL\"\n",
    "    # __version__ = \"1.0\"\n",
    "    # __maintainer__ = \"Will Nash\"\n",
    "    # __email__ = \"will.nash@earlham.ac.uk\"\n",
    "    # __status__ = \"Testing\"\n",
    "    #__modified_by__ \"Sam Witham\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "file_names = 'non-overlapping_includingbidirectional_all_genes_newannotation'\n",
    "Czechowski_rankedcv = '../../data/genes/AtGE_dev_gcRMA__all_probes__CV.tsv'\n",
    "Mergner_rankedcv = '../../data/genes/RNA_CVs.csv'\n",
    "promoter_bedfile = f'../../data/output/{file_names}/FIMO/promoters_5UTR.bed'\n",
    "Czechowski_gene_categories = f'../../data/output/{file_names}/genes/promoters_5UTR_czechowski_constitutive_variable_random.txt'\n",
    "Mergner_gene_categories = f'../../data/output/{file_names}/genes/promoters_5UTR_mergner_constitutive_variable_random.txt'\n",
    "no_of_genes = 100\n",
    "promoter_mapped_motifs = f'../../data/output/{file_names}/FIMO/promoters_5UTR_motifs_mapped.bed'\n",
    "promoters_filtered_contain_motifs = f'../../data/output/{file_names}/FIMO/promoters_5UTR_filtered_contain_motifs.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory  ../../data/output/non-overlapping_includingbidirectional_all_genes_newannotation/genes  already exists\n"
     ]
    }
   ],
   "source": [
    "#make directory for the output files to be exported to\n",
    "#dirName = f'{args.directory_path}/data/output/{args.file_names}'\n",
    "dirName = f'../../data/output/{file_names}/genes'\n",
    "try:\n",
    "    # Create target Directory\n",
    "    os.mkdir(dirName)\n",
    "    print(\"Directory \" , dirName ,  \" created\") \n",
    "except FileExistsError:\n",
    "    print(\"Directory \" , dirName ,  \" already exists\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "remove_proms_no_TFBS(promoter_bedfile, promoter_mapped_motifs,promoters_filtered_contain_motifs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>AGI</th>\n",
       "      <th>dot1</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>attributes</th>\n",
       "      <th>rank</th>\n",
       "      <th>probe_id</th>\n",
       "      <th>expression_mean</th>\n",
       "      <th>expression_SD</th>\n",
       "      <th>expression_CV</th>\n",
       "      <th>proportion_of_values_present_in_mas5</th>\n",
       "      <th>presence_in_araport11</th>\n",
       "      <th>constitutive_in_araport11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>37061</td>\n",
       "      <td>38443</td>\n",
       "      <td>AT1G01060</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT1G01060;Name=LHY;biotype=protein_cod...</td>\n",
       "      <td>8752.0</td>\n",
       "      <td>261569_at</td>\n",
       "      <td>659.659794</td>\n",
       "      <td>490.970711</td>\n",
       "      <td>0.744279</td>\n",
       "      <td>99.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>46789</td>\n",
       "      <td>47233</td>\n",
       "      <td>AT1G01080</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT1G01080;biotype=protein_coding;descr...</td>\n",
       "      <td>8901.0</td>\n",
       "      <td>261577_at</td>\n",
       "      <td>384.087612</td>\n",
       "      <td>292.389851</td>\n",
       "      <td>0.761258</td>\n",
       "      <td>96.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>49166</td>\n",
       "      <td>49908</td>\n",
       "      <td>AT1G01090</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT1G01090;Name=PDH-E1 ALPHA;biotype=pr...</td>\n",
       "      <td>7123.0</td>\n",
       "      <td>261583_at</td>\n",
       "      <td>2541.430002</td>\n",
       "      <td>1529.913875</td>\n",
       "      <td>0.601989</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>50954</td>\n",
       "      <td>51952</td>\n",
       "      <td>AT1G01100</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT1G01100;Name=RPP1A;biotype=protein_c...</td>\n",
       "      <td>4823.0</td>\n",
       "      <td>261578_at</td>\n",
       "      <td>11304.396095</td>\n",
       "      <td>5166.610692</td>\n",
       "      <td>0.457044</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>58978</td>\n",
       "      <td>60215</td>\n",
       "      <td>AT1G01120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT1G01120;Name=KCS1;biotype=protein_co...</td>\n",
       "      <td>6864.0</td>\n",
       "      <td>261570_at</td>\n",
       "      <td>991.741578</td>\n",
       "      <td>578.774828</td>\n",
       "      <td>0.583594</td>\n",
       "      <td>96.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19568</th>\n",
       "      <td>5</td>\n",
       "      <td>26923274</td>\n",
       "      <td>26924488</td>\n",
       "      <td>AT5G67460</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT5G67460;biotype=protein_coding;descr...</td>\n",
       "      <td>10771.0</td>\n",
       "      <td>246994_at</td>\n",
       "      <td>143.437529</td>\n",
       "      <td>158.091065</td>\n",
       "      <td>1.102160</td>\n",
       "      <td>94.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19569</th>\n",
       "      <td>5</td>\n",
       "      <td>26934694</td>\n",
       "      <td>26935074</td>\n",
       "      <td>AT5G67490</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT5G67490;biotype=protein_coding;descr...</td>\n",
       "      <td>5261.0</td>\n",
       "      <td>247006_at</td>\n",
       "      <td>937.252790</td>\n",
       "      <td>449.548526</td>\n",
       "      <td>0.479645</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19572</th>\n",
       "      <td>5</td>\n",
       "      <td>26949184</td>\n",
       "      <td>26950578</td>\n",
       "      <td>AT5G67560</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT5G67560;Name=ARL8B;biotype=protein_c...</td>\n",
       "      <td>3814.0</td>\n",
       "      <td>247008_at</td>\n",
       "      <td>711.205117</td>\n",
       "      <td>287.266875</td>\n",
       "      <td>0.403916</td>\n",
       "      <td>97.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19573</th>\n",
       "      <td>5</td>\n",
       "      <td>26957073</td>\n",
       "      <td>26957914</td>\n",
       "      <td>AT5G67580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT5G67580;Name=TRB2;biotype=protein_co...</td>\n",
       "      <td>206.0</td>\n",
       "      <td>247012_at</td>\n",
       "      <td>61.548600</td>\n",
       "      <td>13.354799</td>\n",
       "      <td>0.216980</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19576</th>\n",
       "      <td>5</td>\n",
       "      <td>26969306</td>\n",
       "      <td>26969515</td>\n",
       "      <td>AT5G67630</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT5G67630;biotype=protein_coding;descr...</td>\n",
       "      <td>5449.0</td>\n",
       "      <td>247014_at</td>\n",
       "      <td>112.439116</td>\n",
       "      <td>55.244540</td>\n",
       "      <td>0.491328</td>\n",
       "      <td>98.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7794 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr     start      stop        AGI dot1 strand     source      type  \\\n",
       "3        1     37061     38443  AT1G01060    .      -  araport11  promoter   \n",
       "5        1     46789     47233  AT1G01080    .      -  araport11  promoter   \n",
       "6        1     49166     49908  AT1G01090    .      -  araport11  promoter   \n",
       "7        1     50954     51952  AT1G01100    .      -  araport11  promoter   \n",
       "9        1     58978     60215  AT1G01120    .      -  araport11  promoter   \n",
       "...    ...       ...       ...        ...  ...    ...        ...       ...   \n",
       "19568    5  26923274  26924488  AT5G67460    .      -  araport11  promoter   \n",
       "19569    5  26934694  26935074  AT5G67490    .      -  araport11  promoter   \n",
       "19572    5  26949184  26950578  AT5G67560    .      +  araport11  promoter   \n",
       "19573    5  26957073  26957914  AT5G67580    .      -  araport11  promoter   \n",
       "19576    5  26969306  26969515  AT5G67630    .      -  araport11  promoter   \n",
       "\n",
       "      dot2                                         attributes     rank  \\\n",
       "3        .  ID=gene:AT1G01060;Name=LHY;biotype=protein_cod...   8752.0   \n",
       "5        .  ID=gene:AT1G01080;biotype=protein_coding;descr...   8901.0   \n",
       "6        .  ID=gene:AT1G01090;Name=PDH-E1 ALPHA;biotype=pr...   7123.0   \n",
       "7        .  ID=gene:AT1G01100;Name=RPP1A;biotype=protein_c...   4823.0   \n",
       "9        .  ID=gene:AT1G01120;Name=KCS1;biotype=protein_co...   6864.0   \n",
       "...    ...                                                ...      ...   \n",
       "19568    .  ID=gene:AT5G67460;biotype=protein_coding;descr...  10771.0   \n",
       "19569    .  ID=gene:AT5G67490;biotype=protein_coding;descr...   5261.0   \n",
       "19572    .  ID=gene:AT5G67560;Name=ARL8B;biotype=protein_c...   3814.0   \n",
       "19573    .  ID=gene:AT5G67580;Name=TRB2;biotype=protein_co...    206.0   \n",
       "19576    .  ID=gene:AT5G67630;biotype=protein_coding;descr...   5449.0   \n",
       "\n",
       "        probe_id  expression_mean  expression_SD  expression_CV  \\\n",
       "3      261569_at       659.659794     490.970711       0.744279   \n",
       "5      261577_at       384.087612     292.389851       0.761258   \n",
       "6      261583_at      2541.430002    1529.913875       0.601989   \n",
       "7      261578_at     11304.396095    5166.610692       0.457044   \n",
       "9      261570_at       991.741578     578.774828       0.583594   \n",
       "...          ...              ...            ...            ...   \n",
       "19568  246994_at       143.437529     158.091065       1.102160   \n",
       "19569  247006_at       937.252790     449.548526       0.479645   \n",
       "19572  247008_at       711.205117     287.266875       0.403916   \n",
       "19573  247012_at        61.548600      13.354799       0.216980   \n",
       "19576  247014_at       112.439116      55.244540       0.491328   \n",
       "\n",
       "       proportion_of_values_present_in_mas5  presence_in_araport11  \\\n",
       "3                                      99.0                    1.0   \n",
       "5                                      96.0                    1.0   \n",
       "6                                     100.0                    1.0   \n",
       "7                                     100.0                    1.0   \n",
       "9                                      96.0                    1.0   \n",
       "...                                     ...                    ...   \n",
       "19568                                  94.0                    1.0   \n",
       "19569                                 100.0                    1.0   \n",
       "19572                                  97.0                    1.0   \n",
       "19573                                 100.0                    1.0   \n",
       "19576                                  98.0                    1.0   \n",
       "\n",
       "       constitutive_in_araport11  \n",
       "3                            0.0  \n",
       "5                            0.0  \n",
       "6                            0.0  \n",
       "7                            0.0  \n",
       "9                            0.0  \n",
       "...                          ...  \n",
       "19568                        0.0  \n",
       "19569                        0.0  \n",
       "19572                        0.0  \n",
       "19573                        0.0  \n",
       "19576                        0.0  \n",
       "\n",
       "[7794 rows x 18 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_czechowski = filter_genes_czechowski(promoters_filtered_no_motifs,Czechowski_rankedcv)\n",
    "filtered_czechowski"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>AGI</th>\n",
       "      <th>dot1</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcription_class</th>\n",
       "      <th>transcription_family</th>\n",
       "      <th>expression_CV</th>\n",
       "      <th>presence_in_araport11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>2630</td>\n",
       "      <td>3759</td>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT1G01010;Name=NAC001;biotype=protein_...</td>\n",
       "      <td>TF</td>\n",
       "      <td>NAC</td>\n",
       "      <td>0.945805</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>8666</td>\n",
       "      <td>10130</td>\n",
       "      <td>AT1G01020</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT1G01020;Name=ARV1;biotype=protein_co...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.384746</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>12940</td>\n",
       "      <td>14714</td>\n",
       "      <td>AT1G01030</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT1G01030;Name=NGA3;biotype=protein_co...</td>\n",
       "      <td>TF</td>\n",
       "      <td>B3</td>\n",
       "      <td>1.094770</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>37061</td>\n",
       "      <td>38443</td>\n",
       "      <td>AT1G01060</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT1G01060;Name=LHY;biotype=protein_cod...</td>\n",
       "      <td>TF</td>\n",
       "      <td>MYB-related</td>\n",
       "      <td>0.877298</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>40877</td>\n",
       "      <td>42017</td>\n",
       "      <td>AT1G01070</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT1G01070;biotype=protein_coding;descr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.202535</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19573</th>\n",
       "      <td>5</td>\n",
       "      <td>26957073</td>\n",
       "      <td>26957914</td>\n",
       "      <td>AT5G67580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT5G67580;Name=TRB2;biotype=protein_co...</td>\n",
       "      <td>TF</td>\n",
       "      <td>MYB-related</td>\n",
       "      <td>1.039561</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19574</th>\n",
       "      <td>5</td>\n",
       "      <td>26963614</td>\n",
       "      <td>26964550</td>\n",
       "      <td>AT5G67610</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT5G67610;biotype=protein_coding;descr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.347671</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19575</th>\n",
       "      <td>5</td>\n",
       "      <td>26965720</td>\n",
       "      <td>26967010</td>\n",
       "      <td>AT5G67620</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT5G67620;biotype=protein_coding;descr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.288362</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19576</th>\n",
       "      <td>5</td>\n",
       "      <td>26969306</td>\n",
       "      <td>26969515</td>\n",
       "      <td>AT5G67630</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT5G67630;biotype=protein_coding;descr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.927208</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19577</th>\n",
       "      <td>5</td>\n",
       "      <td>26970548</td>\n",
       "      <td>26971668</td>\n",
       "      <td>AT5G67640</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>araport11</td>\n",
       "      <td>promoter</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:AT5G67640;biotype=protein_coding;descr...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.386043</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17647 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr     start      stop        AGI dot1 strand     source      type  \\\n",
       "0        1      2630      3759  AT1G01010    .      +  araport11  promoter   \n",
       "1        1      8666     10130  AT1G01020    .      -  araport11  promoter   \n",
       "2        1     12940     14714  AT1G01030    .      -  araport11  promoter   \n",
       "3        1     37061     38443  AT1G01060    .      -  araport11  promoter   \n",
       "4        1     40877     42017  AT1G01070    .      -  araport11  promoter   \n",
       "...    ...       ...       ...        ...  ...    ...        ...       ...   \n",
       "19573    5  26957073  26957914  AT5G67580    .      -  araport11  promoter   \n",
       "19574    5  26963614  26964550  AT5G67610    .      -  araport11  promoter   \n",
       "19575    5  26965720  26967010  AT5G67620    .      -  araport11  promoter   \n",
       "19576    5  26969306  26969515  AT5G67630    .      -  araport11  promoter   \n",
       "19577    5  26970548  26971668  AT5G67640    .      -  araport11  promoter   \n",
       "\n",
       "      dot2                                         attributes  \\\n",
       "0        .  ID=gene:AT1G01010;Name=NAC001;biotype=protein_...   \n",
       "1        .  ID=gene:AT1G01020;Name=ARV1;biotype=protein_co...   \n",
       "2        .  ID=gene:AT1G01030;Name=NGA3;biotype=protein_co...   \n",
       "3        .  ID=gene:AT1G01060;Name=LHY;biotype=protein_cod...   \n",
       "4        .  ID=gene:AT1G01070;biotype=protein_coding;descr...   \n",
       "...    ...                                                ...   \n",
       "19573    .  ID=gene:AT5G67580;Name=TRB2;biotype=protein_co...   \n",
       "19574    .  ID=gene:AT5G67610;biotype=protein_coding;descr...   \n",
       "19575    .  ID=gene:AT5G67620;biotype=protein_coding;descr...   \n",
       "19576    .  ID=gene:AT5G67630;biotype=protein_coding;descr...   \n",
       "19577    .  ID=gene:AT5G67640;biotype=protein_coding;descr...   \n",
       "\n",
       "      transcription_class transcription_family  expression_CV  \\\n",
       "0                      TF                  NAC       0.945805   \n",
       "1                     NaN                  NaN       0.384746   \n",
       "2                      TF                   B3       1.094770   \n",
       "3                      TF          MYB-related       0.877298   \n",
       "4                     NaN                  NaN       1.202535   \n",
       "...                   ...                  ...            ...   \n",
       "19573                  TF          MYB-related       1.039561   \n",
       "19574                 NaN                  NaN       0.347671   \n",
       "19575                 NaN                  NaN       1.288362   \n",
       "19576                 NaN                  NaN       0.927208   \n",
       "19577                 NaN                  NaN       0.386043   \n",
       "\n",
       "       presence_in_araport11  \n",
       "0                        1.0  \n",
       "1                        1.0  \n",
       "2                        1.0  \n",
       "3                        1.0  \n",
       "4                        1.0  \n",
       "...                      ...  \n",
       "19573                    1.0  \n",
       "19574                    1.0  \n",
       "19575                    1.0  \n",
       "19576                    1.0  \n",
       "19577                    1.0  \n",
       "\n",
       "[17647 rows x 14 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_mergner = filter_genes_mergner(promoters_filtered_no_motifs,Mergner_rankedcv)\n",
    "filtered_mergner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#czechowksi subset\n",
    "subSet_onCV(filtered_czechowski,Czechowski_gene_categories,no_of_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mergner subset\n",
    "subSet_onCV(filtered_mergner,Mergner_gene_categories,no_of_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:PromoterArchitecturePipeline] *",
   "language": "python",
   "name": "conda-env-PromoterArchitecturePipeline-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
