{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pybedtools import BedTool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "select_genes_file = '../../data/genomes/ara_housekeeping_list.out'\n",
    "select_genes_file_100_random = '../../data/genes/constitutive-variable-random_100_each.csv'\n",
    "genes_overlap_bedfile = '../../data/genomes/promoterandgenes_only_overlap.bed'\n",
    "#this is actually genome features overlap\n",
    "promoters_overlapping_bedfile = '../../data/genomes/promoters_overlapping.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "select_genes = pd.read_table(select_genes_file, sep='\\t', header=None)\n",
    "cols = ['gene','gene_type']\n",
    "select_genes.columns = cols\n",
    "genes_overlap = pd.read_table(genes_overlap_bedfile, sep='\\t', header=None)\n",
    "cols = ['chr', 'start', 'stop', 'gene', 'dot', 'strand', 'source', 'type', 'dot2', 'details']\n",
    "genes_overlap.columns = cols\n",
    "promoters_overlapping = pd.read_table(promoters_overlapping_bedfile, sep='\\t', header=None)\n",
    "promoters_overlapping.columns = cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G34270</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT3G32260</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G59830</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT4G33380</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT2G28390</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>AT5G13410</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>AT5G56090</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>AT5G05170</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>AT3G18110</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>AT3G13445</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene     gene_type\n",
       "0    AT4G34270  housekeeping\n",
       "1    AT3G32260  housekeeping\n",
       "2    AT1G59830  housekeeping\n",
       "3    AT4G33380  housekeeping\n",
       "4    AT2G28390  housekeeping\n",
       "..         ...           ...\n",
       "295  AT5G13410      randCont\n",
       "296  AT5G56090      randCont\n",
       "297  AT5G05170      randCont\n",
       "298  AT3G18110      randCont\n",
       "299  AT3G13445      randCont\n",
       "\n",
       "[300 rows x 2 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "random_100_only = pd.read_csv(select_genes_file_100_random, header=0)\n",
    "del random_100_only['Unnamed: 0']\n",
    "random_100_only.rename(columns={'promoter_AGI':'gene'}, inplace=True)\n",
    "random_100_only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G34270</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT3G32260</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G59830</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT4G33380</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT2G28390</td>\n",
       "      <td>housekeeping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401</th>\n",
       "      <td>AT1G65490</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>402</th>\n",
       "      <td>AT2G30360</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>403</th>\n",
       "      <td>AT5G56270</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>AT5G57785</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>AT2G46240</td>\n",
       "      <td>randCont</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>406 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene     gene_type\n",
       "0    AT4G34270  housekeeping\n",
       "1    AT3G32260  housekeeping\n",
       "2    AT1G59830  housekeeping\n",
       "3    AT4G33380  housekeeping\n",
       "4    AT2G28390  housekeeping\n",
       "..         ...           ...\n",
       "401  AT1G65490      randCont\n",
       "402  AT2G30360      randCont\n",
       "403  AT5G56270      randCont\n",
       "404  AT5G57785      randCont\n",
       "405  AT2G46240      randCont\n",
       "\n",
       "[406 rows x 2 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "select_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### filter out unwanted genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = pd.merge(select_genes, genes_overlap, on='gene')\n",
    "merged2 = pd.merge(select_genes, promoters_overlapping, on='gene')\n",
    "merged_random100 = pd.merge(random_100_only, genes_overlap, on='gene')\n",
    "merged2_random100 = pd.merge(random_100_only, promoters_overlapping, on='gene')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>dot</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G34270</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>Chr4</td>\n",
       "      <td>16406303</td>\n",
       "      <td>16407303</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:8893552,PMID:16166256,PMID:1708550...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G59830</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>22022503</td>\n",
       "      <td>22023503</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:10091592,PMID:10080713,PMID:986939...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT4G33380</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>Chr4</td>\n",
       "      <td>16070739</td>\n",
       "      <td>16071739</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:16166256,PMID:17085508,PMID:179083...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT5G55840</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>Chr5</td>\n",
       "      <td>22596992</td>\n",
       "      <td>22597992</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15632092,PMID:15565108,PMID:161662...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G13320</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>4567849</td>\n",
       "      <td>4568849</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:7811971,PMID:14973165,PMID:1527645...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>AT1G27370</td>\n",
       "      <td>randCont</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>9508542</td>\n",
       "      <td>9509542</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:11118137,PMID:14972688,PMID:151310...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>AT1G10070</td>\n",
       "      <td>randCont</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>3287086</td>\n",
       "      <td>3288086</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ATBCAT-2,branched-chain amino acid trans...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>AT1G65490</td>\n",
       "      <td>randCont</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>24353645</td>\n",
       "      <td>24354645</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15894741,PMID:18650403,PMID:207364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>AT2G30360</td>\n",
       "      <td>randCont</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>12938834</td>\n",
       "      <td>12939834</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=CIPK11,CBL-INTERACTING PROTEIN KINASE 11...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>AT2G46240</td>\n",
       "      <td>randCont</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>18985132</td>\n",
       "      <td>18986132</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ATBAG6,ARABIDOPSIS THALIANA BCL-2-ASSOCI...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>254 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene     gene_type   chr     start      stop dot strand     source  \\\n",
       "0    AT4G34270  housekeeping  Chr4  16406303  16407303   .      -  Araport11   \n",
       "1    AT1G59830  housekeeping  Chr1  22022503  22023503   .      -  Araport11   \n",
       "2    AT4G33380  housekeeping  Chr4  16070739  16071739   .      +  Araport11   \n",
       "3    AT5G55840  housekeeping  Chr5  22596992  22597992   .      +  Araport11   \n",
       "4    AT1G13320  housekeeping  Chr1   4567849   4568849   .      -  Araport11   \n",
       "..         ...           ...   ...       ...       ...  ..    ...        ...   \n",
       "249  AT1G27370      randCont  Chr1   9508542   9509542   .      -  Araport11   \n",
       "250  AT1G10070      randCont  Chr1   3287086   3288086   .      +  Araport11   \n",
       "251  AT1G65490      randCont  Chr1  24353645  24354645   .      +  Araport11   \n",
       "252  AT2G30360      randCont  Chr2  12938834  12939834   .      -  Araport11   \n",
       "253  AT2G46240      randCont  Chr2  18985132  18986132   .      +  Araport11   \n",
       "\n",
       "     type dot2                                            details  \n",
       "0    gene    .  Dbxref=PMID:8893552,PMID:16166256,PMID:1708550...  \n",
       "1    gene    .  Dbxref=PMID:10091592,PMID:10080713,PMID:986939...  \n",
       "2    gene    .  Dbxref=PMID:16166256,PMID:17085508,PMID:179083...  \n",
       "3    gene    .  Dbxref=PMID:15632092,PMID:15565108,PMID:161662...  \n",
       "4    gene    .  Dbxref=PMID:7811971,PMID:14973165,PMID:1527645...  \n",
       "..    ...  ...                                                ...  \n",
       "249  gene    .  Dbxref=PMID:11118137,PMID:14972688,PMID:151310...  \n",
       "250  gene    .  Alias=ATBCAT-2,branched-chain amino acid trans...  \n",
       "251  gene    .  Dbxref=PMID:15894741,PMID:18650403,PMID:207364...  \n",
       "252  gene    .  Alias=CIPK11,CBL-INTERACTING PROTEIN KINASE 11...  \n",
       "253  gene    .  Alias=ATBAG6,ARABIDOPSIS THALIANA BCL-2-ASSOCI...  \n",
       "\n",
       "[254 rows x 11 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "housekeeping    66\n",
       "randCont        59\n",
       "highVar         56\n",
       "Name: gene_type, dtype: int64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## how many of each gene_type, genes_overlap\n",
    "merged_counts = merged_random100['gene_type'].value_counts()\n",
    "merged_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "housekeeping    85\n",
       "highVar         73\n",
       "randCont        73\n",
       "Name: gene_type, dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## how many of each gene_type, genome features overlap\n",
    "merged_counts2 = merged2_random100['gene_type'].value_counts()\n",
    "merged_counts2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>dot</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G34270</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>Chr4</td>\n",
       "      <td>16406303</td>\n",
       "      <td>16407303</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:8893552,PMID:16166256,PMID:1708550...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G59830</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>22022503</td>\n",
       "      <td>22023503</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:10091592,PMID:10080713,PMID:986939...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT4G33380</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>Chr4</td>\n",
       "      <td>16070739</td>\n",
       "      <td>16071739</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:16166256,PMID:17085508,PMID:179083...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT2G28390</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>12143665</td>\n",
       "      <td>12144665</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=MON1,MONENSIN SENSITIVITY1;Dbxref=PMID:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT5G55840</td>\n",
       "      <td>housekeeping</td>\n",
       "      <td>Chr5</td>\n",
       "      <td>22596992</td>\n",
       "      <td>22597992</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15632092,PMID:15565108,PMID:161662...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>AT3G43960</td>\n",
       "      <td>randCont</td>\n",
       "      <td>Chr3</td>\n",
       "      <td>15775873</td>\n",
       "      <td>15776873</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:12068095,PMID:12805588,PMID:146972...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>AT1G65490</td>\n",
       "      <td>randCont</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>24353645</td>\n",
       "      <td>24354645</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15894741,PMID:18650403,PMID:207364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>AT2G30360</td>\n",
       "      <td>randCont</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>12938834</td>\n",
       "      <td>12939834</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=CIPK11,CBL-INTERACTING PROTEIN KINASE 11...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>AT5G56270</td>\n",
       "      <td>randCont</td>\n",
       "      <td>Chr5</td>\n",
       "      <td>22778613</td>\n",
       "      <td>22779613</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ATWRKY2,ARABIDOPSIS THALIANA WRKY DNA-BI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>AT2G46240</td>\n",
       "      <td>randCont</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>18985132</td>\n",
       "      <td>18986132</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ATBAG6,ARABIDOPSIS THALIANA BCL-2-ASSOCI...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>320 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene     gene_type   chr     start      stop dot strand     source  \\\n",
       "0    AT4G34270  housekeeping  Chr4  16406303  16407303   .      -  Araport11   \n",
       "1    AT1G59830  housekeeping  Chr1  22022503  22023503   .      -  Araport11   \n",
       "2    AT4G33380  housekeeping  Chr4  16070739  16071739   .      +  Araport11   \n",
       "3    AT2G28390  housekeeping  Chr2  12143665  12144665   .      -  Araport11   \n",
       "4    AT5G55840  housekeeping  Chr5  22596992  22597992   .      +  Araport11   \n",
       "..         ...           ...   ...       ...       ...  ..    ...        ...   \n",
       "315  AT3G43960      randCont  Chr3  15775873  15776873   .      -  Araport11   \n",
       "316  AT1G65490      randCont  Chr1  24353645  24354645   .      +  Araport11   \n",
       "317  AT2G30360      randCont  Chr2  12938834  12939834   .      -  Araport11   \n",
       "318  AT5G56270      randCont  Chr5  22778613  22779613   .      +  Araport11   \n",
       "319  AT2G46240      randCont  Chr2  18985132  18986132   .      +  Araport11   \n",
       "\n",
       "     type dot2                                            details  \n",
       "0    gene    .  Dbxref=PMID:8893552,PMID:16166256,PMID:1708550...  \n",
       "1    gene    .  Dbxref=PMID:10091592,PMID:10080713,PMID:986939...  \n",
       "2    gene    .  Dbxref=PMID:16166256,PMID:17085508,PMID:179083...  \n",
       "3    gene    .  Alias=MON1,MONENSIN SENSITIVITY1;Dbxref=PMID:1...  \n",
       "4    gene    .  Dbxref=PMID:15632092,PMID:15565108,PMID:161662...  \n",
       "..    ...  ...                                                ...  \n",
       "315  gene    .  Dbxref=PMID:12068095,PMID:12805588,PMID:146972...  \n",
       "316  gene    .  Dbxref=PMID:15894741,PMID:18650403,PMID:207364...  \n",
       "317  gene    .  Alias=CIPK11,CBL-INTERACTING PROTEIN KINASE 11...  \n",
       "318  gene    .  Alias=ATWRKY2,ARABIDOPSIS THALIANA WRKY DNA-BI...  \n",
       "319  gene    .  Alias=ATBAG6,ARABIDOPSIS THALIANA BCL-2-ASSOCI...  \n",
       "\n",
       "[320 rows x 11 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>dot</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene, gene_type, chr, start, stop, dot, strand, source, type, dot2, details]\n",
       "Index: []"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check if any NaN - there won't be NaN as did inner merge\n",
    "merged[pd.isna(merged.chr)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### change column order to bed file format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_proms = merged.loc[:, ['chr', 'start', 'stop', 'gene', 'dot', 'strand', 'source', 'type', 'dot2', 'details']]\n",
    "filtered_proms2 = merged2.loc[:, ['chr', 'start', 'stop', 'gene', 'dot', 'strand', 'source', 'type', 'dot2', 'details']]\n",
    "#sort the df by chromosome then by motif start position\n",
    "filtered_proms_sorted = filtered_proms.sort_values(['chr','start'])\n",
    "filtered_proms_sorted2 = filtered_proms2.sort_values(['chr','start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>gene</th>\n",
       "      <th>dot</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>dot2</th>\n",
       "      <th>details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>33171</td>\n",
       "      <td>34171</td>\n",
       "      <td>AT1G01050</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=AtPPa1,pyrophosphorylase 1;Dbxref=PMID:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>281760</td>\n",
       "      <td>282760</td>\n",
       "      <td>AT1G01780</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:15642518,PMID:15659355,PMID:155317...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>316090</td>\n",
       "      <td>317090</td>\n",
       "      <td>AT1G01910</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:18315867,PMID:18775970,PMID:207364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>844251</td>\n",
       "      <td>845251</td>\n",
       "      <td>AT1G03400</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:7579161,PMID:12805619,PMID:1761651...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>892492</td>\n",
       "      <td>893492</td>\n",
       "      <td>AT1G03560</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:16709272,PMID:16951057,locus:20208...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>25155983</td>\n",
       "      <td>25156983</td>\n",
       "      <td>AT5G62650</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=locus:2172234;ID=AT5G62650;Name=AT5G626...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>25633209</td>\n",
       "      <td>25634209</td>\n",
       "      <td>AT5G64050</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Alias=ATERS,OVA3,OVULE ABORTION 3;Dbxref=PMID:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>25909014</td>\n",
       "      <td>25910014</td>\n",
       "      <td>AT5G64813</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:17683937,PMID:23144185,locus:50500...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>26467531</td>\n",
       "      <td>26468531</td>\n",
       "      <td>AT5G66250</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:18650403,PMID:18775970,PMID:207364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>Chr5</td>\n",
       "      <td>26956914</td>\n",
       "      <td>26957914</td>\n",
       "      <td>AT5G67590</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>.</td>\n",
       "      <td>Dbxref=PMID:12084824,PMID:12837548,PMID:146710...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>254 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr     start      stop       gene dot strand     source  type dot2  \\\n",
       "144  Chr1     33171     34171  AT1G01050   .      -  Araport11  gene    .   \n",
       "94   Chr1    281760    282760  AT1G01780   .      +  Araport11  gene    .   \n",
       "32   Chr1    316090    317090  AT1G01910   .      -  Araport11  gene    .   \n",
       "238  Chr1    844251    845251  AT1G03400   .      -  Araport11  gene    .   \n",
       "134  Chr1    892492    893492  AT1G03560   .      -  Araport11  gene    .   \n",
       "..    ...       ...       ...        ...  ..    ...        ...   ...  ...   \n",
       "173  Chr5  25155983  25156983  AT5G62650   .      -  Araport11  gene    .   \n",
       "205  Chr5  25633209  25634209  AT5G64050   .      -  Araport11  gene    .   \n",
       "27   Chr5  25909014  25910014  AT5G64813   .      +  Araport11  gene    .   \n",
       "230  Chr5  26467531  26468531  AT5G66250   .      +  Araport11  gene    .   \n",
       "129  Chr5  26956914  26957914  AT5G67590   .      +  Araport11  gene    .   \n",
       "\n",
       "                                               details  \n",
       "144  Alias=AtPPa1,pyrophosphorylase 1;Dbxref=PMID:1...  \n",
       "94   Dbxref=PMID:15642518,PMID:15659355,PMID:155317...  \n",
       "32   Dbxref=PMID:18315867,PMID:18775970,PMID:207364...  \n",
       "238  Dbxref=PMID:7579161,PMID:12805619,PMID:1761651...  \n",
       "134  Dbxref=PMID:16709272,PMID:16951057,locus:20208...  \n",
       "..                                                 ...  \n",
       "173  Dbxref=locus:2172234;ID=AT5G62650;Name=AT5G626...  \n",
       "205  Alias=ATERS,OVA3,OVULE ABORTION 3;Dbxref=PMID:...  \n",
       "27   Dbxref=PMID:17683937,PMID:23144185,locus:50500...  \n",
       "230  Dbxref=PMID:18650403,PMID:18775970,PMID:207364...  \n",
       "129  Dbxref=PMID:12084824,PMID:12837548,PMID:146710...  \n",
       "\n",
       "[254 rows x 10 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_proms_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### write out bed file of merged dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(../../data/genomes/promoters_overlapping__const_var.bed)>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BedTool.from_dataframe(filtered_proms_sorted).saveas('../../data/genomes/promoterandgenes_only_overlap_const_var.bed')\n",
    "BedTool.from_dataframe(filtered_proms_sorted2).saveas('../../data/genomes/promoters_overlapping__const_var.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:PromoterArchitecturePipeline] *",
   "language": "python",
   "name": "conda-env-PromoterArchitecturePipeline-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
